### ISSUE TITLE
Return value from OQS_randombytes

### ISSUE THREAD

**Issue by matlimatli** [open-quantum-safe/liboqs#1750]
The current signature of `OQS_Randombytes()` is to return `void`. This might be a limiting factor, which is also noted in `OQS_randombytes_openssl()` with the following comment:
```
// because of void signature we have no other way to signal the problem
// we cannot possibly return without randomness
```
and the failure handling of this function is therefore to terminate the execution via `exit()`.

In cases where we supply our own randomness algorithm via `OQS_randombytes_custom_algorithm()`, the same problem applies. We have no other means of handling a failure than forcefully exiting. It is not unthinkable that the higher level code would want to take care of this situation in a custom way without having to resort to using an exit handler.

Would it be feasible to change this signature to returning a status value, which can then be propagated through the higher level functions? For example, the keypair and encaps functions already return an `OQS_STATUS` value, which can be taken care of in the calling code. Wouldn't it make sense if OQS_ERROR was returned if an underlying randomness call failed?

**Comment by SWilson4** [comment]
I agree that the `void` return is less than ideal. However, I believe this is a legacy of the [DRBG that NIST provided](https://github.com/open-quantum-safe/liboqs/blob/7a6df815bbb218e4490064a1558b09edea410576/src/common/rand/rand_nist.c) for submissions to the PQC standardization process, which returns `void`. As such, almost all of the code in `liboqs` was written with this function signature in mind.

Modifying `keypair` / `encaps` / etc to error-check `randombytes` would require quite a bit of maintenance on our part. We typically pull in this code from upstream sources (e.g., PQClean / pq-crystals) which would (presumably) still be using the NIST-provided `randombytes` API. Hence, we would have to patch the code and re-patch every time we need to pull in an upstream update.

I suppose that it might not be too difficult to change the `OQS_randombytes` signature to return a status code and just ignore it in `keypair` / `encaps` / etc. while allowing developers to pass an error-checked implementation to `OQS_randombytes_custom_algorithm`. This would also enable us to add better error handling in the future.

**Comment by SWilson4** [comment]
It would be great to hear a little more about your use case and if that latter option would be helpful for you. (And if it would be helpful enough, please feel free to submit a PR proposing the change!)

**Comment by matlimatli** [comment]
Thanks for the explanation! It makes sense (but still is a bit unfortunate).

My use cases are mainly hypothetical at this time, but I am thinking of cases where a long-running system may for example temporarily run out of good entropy. In this case I would expect the cryptographic operations to fail, so that the calling code could remedy the situation (the obvious workaround here is to let the custom randombytes function handle it, which may or may not be a good idea) or shut down the system gracefully.

I'm not sure it would be a good idea to let `OQS_randombytes` return a status code if it is not checked by the callers. It might be dangerous if it could return bad randomness without anyone noticing. In cases where we want to use a function with a different signature, it is probably more safe to use a wrapper, similar to `OQS_randombytes_openssl`.

**Comment by baentsch** [comment]
> I'm not sure it would be a good idea to let OQS_randombytes return a status code if it is not checked by the callers. It might be dangerous if it could return bad randomness without anyone noticing.

Absolutely agreed: All functions based on this (in upstream-imported algorithms) must stop working if entropy is bad.

> In cases where we want to use a function with a different signature, it is probably more safe to use a wrapper, similar to OQS_randombytes_openssl.

Also agreed, but maybe thinking in reverse: What about adding such (result-returning) call to the public `liboqs` API and gradually introducing (or patching it) it to "LTS" code, e.g., the ML-* algorithms? The current (void) call could be a wrapper to this (still just exiting on bad retval)?

**Comment by matlimatli** [comment]
> What about adding such (result-returning) call to the public liboqs API and gradually introducing (or patching it) it to "LTS" code, e.g., the ML-* algorithms? The current (void) call could be a wrapper to this (still just exiting on bad retval)?

Ah, yes, I see. That sounds like a good plan!

**Comment by baentsch** [comment]
> Ah, yes, I see. That sounds like a good plan!

PR welcome :-) Time permitting, of course.

**Comment by owaldron** [comment]
Hi! I've been looking into how this might be possible and wanted to document my findings and open up to input. I collected the existing error codes and potential sources of error from the different `OQS_randombytes` codepaths:

For the different `OQS_RAND_alg_system` paths:
- **Windows `BCryptGenRandom`**: has 3 documented output statuses (documented [here](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom#return-value))
- **Windows OS random**:
	- Three different calls all report a success boolean and have corresponding error codes on failure accessed by calling [`GetLastError`](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror):
		- `CryptAcquireContext`: error codes [here](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptacquirecontextw#return-value)
		- `CryptGenRandom`: error codes [here](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptgenrandom#return-value)
		- `CryptReleaseContext`:  error code [here](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptreleasecontext#return-value)
- **Apple's `arc4random`**: Does not exit softly on error
- **Embedded**: Errors out if asked to use system randomness
- **`getentropy`**: Returns -1 on error and writes error codes to [errno](https://man7.org/linux/man-pages/man3/errno.3.html) ([list of codes](https://man7.org/linux/man-pages/man3/getentropy.3.html#ERRORS))
- Reading **`/dev/urandom`**: Standard file reading errors might include
	- Failure to open file
	- Failure to read bytes from file

For the `OQS_RAND_alg_openssl` path:
- `SIZE_T_TO_INT_OR_EXIT` might exit on overflow
- OpenSSL's `RAND_bytes` could return -1,0, or 1 as documented [here](https://docs.openssl.org/3.1/man3/RAND_bytes/#return-values)

For the custom `OQS_randombytes_custom_algorithm` path, a user-defined function could error out in any arbitrary way.

Between all of this it might make sense to return a `OQS_STATUS` as `OQS_ERROR` and `OQS_SUCCESS` as appropriate, but this would lose some description of the error. A more descriptive error code might be unnecessarily complex without merit.
